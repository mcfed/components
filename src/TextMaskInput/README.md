# TextMaskInput 组件

## Usage

```jsx
// antd4
import React from 'react';
import {Form} from 'antd';
import {TextMaskInput} from '@mcfed/components';

const FormItem = Form.Item;

const Demo = () => {
  const [form] = Form.useForm();
  return (
    <Form form={form} {...layout}>
      <FormItem name='phone' label='phone'>
        <TextMaskInput
          maskType='phone'
          // defaultValue={18868832053}
          onChange={e => console.log(e.target.value)}
          placeholder={'请输入'}
        />
      </FormItem>
      <FormItem name='email' label='email'>
        <TextMaskInput maskType='email' />
      </FormItem>
      <FormItem name='date' label='date'>
        <TextMaskInput maskType='date' />
      </FormItem>
    </Form>
  );
};

// antd3
class Demo extends React.Component {
  render() {
    const {getFieldDecorator} = this.props.form;
    return (
      <Form {...layout}>
        <FormItem label='phone'>
          {getFieldDecorator('phone')(
            <TextMaskInput
              maskType='phone'
              // defaultValue={18868832053}
              onChange={e => console.log(e.target.value)}
              placeholder={'请输入'}
            />
          )}
        </FormItem>
        <FormItem label='email'>
          {getFieldDecorator('email')(<TextMaskInput maskType='email' />)}
        </FormItem>
        <FormItem label='date'>
          {getFieldDecorator('date')(<TextMaskInput maskType='date' />)}
        </FormItem>
      </Form>
    );
  }
}

const App = Form.create()(Demo);

export default App;
```

## Options

| 参数     | 说明      | 类型   | 默认值 | 取值范围         |
| -------- | --------- | ------ | ------ | ---------------- |
| maskType | mask 类型 | string | phone  | phone,email,date |
