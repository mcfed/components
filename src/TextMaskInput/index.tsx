import React from 'react';
import MaskedInput from 'react-text-mask';
import {Input} from 'antd';
import masks from './mask';

type MaskTypes = 'phone' | 'email' | 'date' | 'ip';

interface IMaskedInput {
  props: any;
  setRef: Function;
  onBlur: Function;
  onChange: Function;
}

const maskRender = (ref: Function, props: any) => {
  const {
    defaultValue,
    onChange,
    onChangeField,
    mask,
    maskType,
    pipe,
    ...others
  } = props;
  return (
    <Input
      defaultValue={props.defaultValue}
      ref={input => ref(input && input.input)}
      onChange={event => {
        props.onChangeField && props.onChangeField(event);
      }}
      {...others}
    />
  );
};

class NewMaskedInput extends MaskedInput {
  render(this: IMaskedInput) {
    const {render, ...props} = this.props;
    return maskRender(this.setRef, {
      onBlur: this.onBlur,
      onChangeField: this.onChange,
      defaultValue: this.props.value,
      ...props
    });
  }
}

const MaskInput = ({maskType, ...others}: {maskType: MaskTypes}) => {
  //@ts-ignore
  return <NewMaskedInput {...masks[maskType || 'phone']} {...others} />;
};

export default MaskInput;
