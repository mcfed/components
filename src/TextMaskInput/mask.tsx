import emailMask from 'text-mask-addons/dist/emailMask';
// import createAutoCorrectedDatePipe from 'text-mask-addons/dist/createAutoCorrectedDatePipe';
import ipPipe, {createAutoCorrectedDatePipe} from './pipe';
const autoCorrectedDatePipe = createAutoCorrectedDatePipe('mm/dd/yyyy HH:MM');
const phoneMask = [
  1,
  /[3-9]/,
  /\d/,
  /\d/,
  /\d/,
  /\d/,
  /\d/,
  /\d/,
  /\d/,
  /\d/,
  /\d/
];
const dateMask = [/\d/, /\d/, '/', /\d/, /\d/, '/', /\d/, /\d/, /\d/, /\d/];
const ipMask = [
  /\d/,
  /\d/,
  /\d/,
  '.',
  /\d/,
  /\d/,
  /\d/,
  '.',
  /\d/,
  /\d/,
  /\d/,
  '.',
  /\d/,
  /\d/,
  /\d/
];
const masks = {
  phone: {mask: phoneMask},
  email: {mask: emailMask},
  date: {mask: dateMask, pipe: autoCorrectedDatePipe},
  ip: {mask: ipMask, pipe: ipPipe(), keepCharPositions: true}
};

export default masks;
