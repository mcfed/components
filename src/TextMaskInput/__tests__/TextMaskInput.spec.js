import React from 'react';
import {shallow} from 'enzyme';
import {Form} from 'antd';
import TextMaskInput from '../index';

const Mask = props => {
  <Form>
    <FormItem label='email'>
      {props.form.getFieldDecorator('email')(<TextMaskInput {...props} />)}
    </FormItem>
  </Form>;
};

const Demo = Form.create()(Mask);

const setup = props => {
  const wrapper = shallow(<Demo {...props} />);

  return {
    wrapper,
    props
  };
};

describe('快照测试', () => {
  it('全页快照', () => {
    const props = {
      maskType: 'email'
    };
    const {wrapper} = setup(props);
    expect(wrapper).toMatchSnapshot();
  });
});
