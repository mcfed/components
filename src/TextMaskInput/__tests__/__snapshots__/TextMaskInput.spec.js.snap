// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 全页快照 1`] = `
<Mask
  form={
    Object {
      "getFieldDecorator": [Function],
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldProps": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "isSubmitting": [Function],
      "resetFields": [Function],
      "setFields": [Function],
      "setFieldsInitialValue": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
      "validateFieldsAndScroll": [Function],
    }
  }
  maskType="email"
/>
`;
